rmw_cyclonedds_cpp/src/TypeSupport2.hpp

```diff
@@ -0,0 +1,510 @@
+  }
+};
+
+class ROSIDLC_SpanSequenceValueType : public SpanSequenceValueType
+{
+protected:
+  const AnyValueType * m_element_value_type;
+  struct ROSIDLC_SequenceObject
+  {
+    void * data;
+    size_t size;     /*!< The number of valid items in data */
+    size_t capacity; /*!< The number of allocated items in data */
+  };
+
+  const ROSIDLC_SequenceObject * get_value(const void * ptr_to_sequence) const
+  {
+    return static_cast<const ROSIDLC_SequenceObject *>(ptr_to_sequence);
+  }
+
+public:
+  explicit ROSIDLC_SpanSequenceValueType(const AnyValueType * element_value_type)
+  : m_element_value_type(element_value_type)
+  {
+  }
+
+  size_t sizeof_type() const override {return sizeof(ROSIDLC_SequenceObject);}
+  const AnyValueType * element_value_type() const override {return m_element_value_type;}
+  size_t sequence_size(const void * ptr_to_sequence) const override
+  {
+    return get_value(ptr_to_sequence)->size;
+  }
+  const void * sequence_contents(const void * ptr_to_sequence) const final
+  {
+    return get_value(ptr_to_sequence)->data;
+  }
+};
+
+struct PrimitiveValueType : public AnyValueType
+{
+  const ROSIDL_TypeKind m_type_kind;
+
+  explicit constexpr PrimitiveValueType(ROSIDL_TypeKind type_kind)
+  : m_type_kind(type_kind)
+  {
+    assert(type_kind != ROSIDL_TypeKind::STRING);
+    assert(type_kind != ROSIDL_TypeKind::WSTRING);
+    assert(type_kind != ROSIDL_TypeKind::MESSAGE);
+  }
+  ROSIDL_TypeKind type_kind() const {return m_type_kind;}
+  size_t sizeof_type() const final
+  {
+    switch (m_type_kind) {
+      case ROSIDL_TypeKind::FLOAT:
+        return sizeof(float);
+      case ROSIDL_TypeKind::DOUBLE:
+        return sizeof(double);
+      case ROSIDL_TypeKind::LONG_DOUBLE:
+        return sizeof(long double);
+      case ROSIDL_TypeKind::CHAR:
+        return sizeof(char);
+      case ROSIDL_TypeKind::WCHAR:
+        return sizeof(char16_t);
+      case ROSIDL_TypeKind::BOOLEAN:
+        return sizeof(bool);
+      case ROSIDL_TypeKind::OCTET:
+        return sizeof(unsigned char);
+      case ROSIDL_TypeKind::UINT8:
+        return sizeof(uint_least8_t);
+      case ROSIDL_TypeKind::INT8:
+        return sizeof(int_least8_t);
+      case ROSIDL_TypeKind::UINT16:
+        return sizeof(uint_least16_t);
+      case ROSIDL_TypeKind::INT16:
+        return sizeof(int_least16_t);
+      case ROSIDL_TypeKind::UINT32:
+        return sizeof(uint_least32_t);
+      case ROSIDL_TypeKind::INT32:
+        return sizeof(int_least32_t);
+      case ROSIDL_TypeKind::UINT64:
+        return sizeof(uint_least64_t);
+      case ROSIDL_TypeKind::INT64:
+        return sizeof(int_least64_t);
+      case ROSIDL_TypeKind::STRING:
+      case ROSIDL_TypeKind::WSTRING:
+      case ROSIDL_TypeKind::MESSAGE:
+        throw std::runtime_error(
+                "not a primitive value type: " +
+                std::to_string(std::underlying_type_t<ROSIDL_TypeKind>(m_type_kind)));
+    }
+  }
```

Comment:
Compiler warning: `warning: control reaches end of non-void function [-Wreturn-type]`

Reply:
Thank you. Please take a look at https://github.com/ros2/rmw_cyclonedds/pull/77 which should address this issue